firefox-ios/Client/Utils/ContinuationsChecker.swift

```diff
@@ -25,3 +25,54 @@ struct ContinuationsChecker {
         return true
     }
 }
+
+final class SafeContinuation<T>: @unchecked Sendable {
+    private var continuation: UnsafeContinuation<T, Never>?
+    private var checkedContinuation: CheckedContinuation<T, Never>?
+    private var hasResumed = false
+
+    private let lock = NSLock()
+
+    private init(continuation: UnsafeContinuation<T, Never>) {
+        self.continuation = continuation
+    }
+
+    private init(checkedContinuation: CheckedContinuation<T, Never>) {
+        self.checkedContinuation = checkedContinuation
+    }
+
+    func resume(returning value: T) {
+        lock.lock()
+        defer { lock.unlock() }
+
+        guard !hasResumed else {
+            assertionFailure("⚠️ Continuation already resumed.")
+            return
+        }
+
+        hasResumed = true
+        if let c = continuation {
+            c.resume(returning: value)
+            continuation = nil
+        } else if let c = checkedContinuation {
+            c.resume(returning: value)
+            checkedContinuation = nil
```

Comment:
nit; I didn't test it but I think this works? And would make the code easier to read IMO
```suggestion
        if let continuation {
            continuation.resume(returning: value)
            self.continuation = nil
        } else if let checkedContinuation {
            checkedContinuation.resume(returning: value)
            self.checkedContinuation = nil
```